### ISSUE TITLE
More intelligent subflake locking.

### ISSUE THREAD

**Issue by tejing1** [NixOS/nix#6352]
**Is your feature request related to a problem? Please describe.**
`flake.nix`:
```nix
{
  inputs.child.url = "path:./child";
  outputs = {self, child}: {
    inherit child;
  };
}
```
`child/flake.nix`:
```nix
{
  outputs = {self}: {
    foo = 1;
  };
}
```
```
$ nix eval .#child.foo
1
$ # so far, so good
$ sed -ie 's/1/2/' child/flake.nix
$ nix eval .#child.foo
warning: Git tree '/mnt/persist/share/data/tejing/work/tmpflake' is dirty
1
$ # huh, ok, maybe the dirty tree support doesn't extend that far. I guess that's at least kind of understandable
$ git commit -am 'change foo' &>/dev/null
$ nix eval .#child.foo
1
$ # this is super unintuitive. it really should be 2 at this point
$ nix flake update --commit-lock-file  &>/dev/null
$ nix eval .#child.foo
2
$ # finally!
```

**Describe the solution you'd like**
Nix should recognize that the subflake is already locked by the parent flake's rev, and thus doesn't need its hash put into `flake.lock` at all, removing any possibility of a parent and child from different commits interacting. Ideally the dirty tree case should work too.

**Describe alternatives you've considered**
Nix could try to keep the hash in `flake.lock` up to date somehow, but it fundamentally doesn't have a hook into the proper time to do that, since `git commit` doesn't call nix.
The parent flake could invoke the child by `builtins.getFlake`, rather than as an input, but currently all methods of doing that suffer from similar problems, and you wouldn't have access to input overriding functionality.

**Comment by thufschmitt** [comment]
That’s indeed really annoying. I don’t think using the parent’s lock would be a good idea (if only because it’s too much of an overapproximation), but maybe we could be able to specify that some inputs should be systematically refreshed when running a `nix command`. So you’d have

```nix
{
  inputs.child.url = "path:./child";
  inputs.child.autoUpdate = true;
}
```

and any `nix` command would first call the equivalent of `nix flake lock --update-input child`.

**Comment by Kha** [comment]
Some prior discussion (though tracking this in an open issue is probably a good idea): https://github.com/NixOS/nix/issues/3978#issuecomment-952418478
I really don't think path-relative, flake-local inputs should be locked at all, as they are reproducible by virtue of being included in the flake source, and there is no way to even fetch a stale locked version.

**Comment by aameen-tulip** [comment]
> Some prior discussion (though tracking this in an open issue is probably a good idea): [#3978 (comment)](https://github.com/NixOS/nix/issues/3978#issuecomment-952418478) I really don't think path-relative, flake-local inputs should be locked at all, as they are reproducible by virtue of being included in the flake source, and there is no way to even fetch a stale locked version.

This makes sense except for the case of `nix build github:owner/repo?dir=directly-to-a-subflake#bar` in which case we actually do want the lock for the sub-flake.

**Comment by tejing1** [comment]
> This makes sense except for the case of `nix build github:owner/repo?dir=directly-to-a-subflake#bar` in which case we actually do want the lock for the sub-flake.

It feels like you're misunderstanding here. There are 2 meanings of 'lock' at play. In *all* cases, the subflake's lock *file* should be used. The question is whether the hash of the subflake gets recorded in the lockfile of the consuming flake, when the consuming flake is actually stored in the same git commit. In the case you brought up, there is no consuming flake, and thus the kind of locking in question isn't even happening.

**Comment by aameen-tulip** [comment]
Gotcha. Well as someone who's been bitten badly by subflake locking I definitely agree with "make relative subflakes work more better".

The project that I fully can't get them to work with uses "merge and squash" PR as the only option, so relative paths will always blow up unless there's something I misunderstood about how to manage the locks.

**Comment by blaggacao** [comment]
[Inspired](https://github.com/ursi/get-flake), I submitted [this](https://github.com/edolstra/flake-compat/pull/55) work around.

**Comment by roberth** [comment]
https://github.com/NixOS/nix/pull/6530 improves subflake locking, but is currently blocked on evaluation regressions. We've made some progress merging some stable parts of that PR, so perhaps its solution for subflakes could be extracted and merged independently too?

> - [x] Fix/improve subflake handling.

**Comment by mikepurvis** [comment]
Are subflakes as described in the original post workable at all under today's Nix? I tried to adopt that exact structure and it all worked great locally, but then on Hydra it hit evaluation errors:

```
in job ‘thingy.x86_64-linux’:
error: cannot fetch input 'path:./thing?lastModified=1&narHash=sha256-tk7BEyl8BW4OSrT%2fRCAhYBzJx3LQJxPcV1e2MFRl7kM=' because it uses a relative path
```

This seems to be a complete showstopper as far as being able to use multiple flakes in the same repo for anything other than local development.

**Comment by tomberek** [comment]
the  best fix for this at this moment seems to be to use https://github.com/divnix/call-flake or something similar using [call-flake.nix](https://github.com/NixOS/nix/blob/master/src/libexpr/flake/call-flake.nix):
```
{
   inputs.call-flake.url = "github:divnix/call-flake";
   outputs = inputs: (inputs.call-flake ../..).something.something {};
}
```
I'm aware of this issue. Though I'm not clear on the best path to fix it. @roberth has had proposed some lockfile changes that can help fix.

**Comment by roberth** [comment]
> the best fix for this at this moment

This is _not_ a fix, but rather a workaround. For instance, it can't work with `--override-input`.

I think @edolstra has written an experimental fix, but he hasn't opened a viable PR for it yet.

**Comment by rasmus-kirk** [comment]
What's the status on this?

**Comment by roberth** [comment]
> I really don't think path-relative, flake-local inputs should be locked at all, as they are reproducible by virtue of being included in the flake source

Nix 2.26 includes
- https://github.com/NixOS/nix/pull/10089

I don't see anything in this thread that isn't covered by that (despite plenty of room for scope creep 😄)
Some questions about whether the parent lock should be involved, but I think that would be a separate issue if the current solution isn't adequate. The thread in #10089 does mention desyncing of _transitive_ inputs in the parent lock, as those transitive dependencies are still copied, but we could consider that to be a special case of #7730.

So I think this one can be closed.

**Comment by mikepurvis** [comment]
My use case is met with the changes made in #10089. I had a generated flake that included a small metadata component that I wanted to be able to later use overrides against in order to do hash comparisons.

So like, generated timestamp flake X and generated timestamp flake Y (from say, an hour later) would trivially compare as unique if the timestamp itself is part of the content, but by putting it in a subflake, I can override one to the other when comparing, and that tells me if the content itself has changed (and therefore if the new timestamp flake should be committed and shipped).
